osu.Game.Rulesets.Osu/Objects/Drawables/SpinnerSpinHistory.cs

```diff
@@ -0,0 +1,146 @@
+        /// <param name="currentTime">The current time.</param>
+        /// <param name="delta">The delta of the angle moved through since the last report.</param>
+        public void ReportDelta(double currentTime, float delta)
+        {
+            if (delta == 0)
+                return;
+
+            // Importantly, outside of tests the max delta entering here is 180 degrees.
+            // If it wasn't for tests, we could add this line:
+            //
+            // Debug.Assert(Math.Abs(delta) < 180);
+            //
+            // For this to be 101% correct, we need to add the ability for important frames to be
+            // created based on gameplay intrinsics (ie. there should be one frame for any spinner delta 90 < n < 180 degrees).
+            //
+            // But this can come later.
+
+            totalAbsoluteRotation += delta;
+
+            if (currentTime >= lastReportTime)
+            {
+                currentSpinMaxRotation = Math.Max(currentSpinMaxRotation, Math.Abs(currentSpinRotation));
+
+                // Handle the case where the user has completed another spin.
+                // Note that this does could be an `if` rather than `while` if the above assertion held true.
+                // It is a `while` loop to handle tests which throw larger values at this method.
+                while (currentSpinMaxRotation >= 360)
+                {
+                    int direction = Math.Sign(currentSpinRotation);
+
+                    completedSpins.Push(new CompletedSpin(currentTime, direction));
+
+                    // Incrementing the last completion point will cause `currentSpinRotation` to
+                    // hold the remaining spin that needs to be considered.
+                    totalAbsoluteRotationAtLastCompletion += direction * 360;
+
+                    // Reset the current max as we are entering a new spin.
+                    // Importantly, carry over the remainder (which is now stored in `currentSpinRotation`).
+                    currentSpinMaxRotation = Math.Abs(currentSpinRotation);
+                }
+            }
+            else
+            {
+                // When rewinding, the main thing we care about is getting `totalAbsoluteRotationsAtLastCompletion`
+                // to the correct value. We can used the stored history for this.
+                while (completedSpins.TryPeek(out var segment) && segment.CompletionTime > currentTime)
+                {
+                    completedSpins.Pop();
+                    totalAbsoluteRotationAtLastCompletion -= segment.Direction * 360;
+                }
+
+                // This is a best effort. We may not have enough data to match this 1:1, but that's okay.
+                // We know that the player is somewhere in a spin.
+                // In the worst case, this will be lower than expected, and recover in forward playback.
+                currentSpinMaxRotation = Math.Abs(currentSpinRotation);
```

Comment:
This doesn't actually recover safely as far as I can tell. To demonstrate, here's an adjustment to a test I've commented on above:

```diff
diff --git a/osu.Game.Rulesets.Osu.Tests/SpinnerSpinHistoryTest.cs b/osu.Game.Rulesets.Osu.Tests/SpinnerSpinHistoryTest.cs
index c20a7bbfd9..6452255110 100644
--- a/osu.Game.Rulesets.Osu.Tests/SpinnerSpinHistoryTest.cs
+++ b/osu.Game.Rulesets.Osu.Tests/SpinnerSpinHistoryTest.cs
@@ -129,6 +129,15 @@ public void TestRewindIntoSegmentThatHasNotCrossedZero()
             Assert.That(history.TotalRotation, Is.EqualTo(360));
             history.ReportDelta(1750, -45);
             Assert.That(history.TotalRotation, Is.EqualTo(315));
+
+            // This is equivalent to the t=2000 delta from above, just split into two segments to demonstrate the failure.
+            history.ReportDelta(1800, 50);
+            Assert.That(history.TotalRotation, Is.Not.EqualTo(365)); // We should not have crossed a full spin at this point.
+            Assert.That(history.TotalRotation, Is.EqualTo(180)); // Expected value - max accumulated rotation within one spin.
+
+            history.ReportDelta(2000, 400);
+            Assert.That(history.TotalRotation, Is.Not.EqualTo(765)); // We should not have 2 full spins.
+            Assert.That(history.TotalRotation, Is.EqualTo(360)); // Expected value - exactly one full spin.
         }

         [Test]
```

In other words, the worst case is actually that you have extra spins, not fewer spins.